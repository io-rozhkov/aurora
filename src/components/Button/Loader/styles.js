import styled, { keyframes } from "styled-components";

import { getThemeValue } from "../../../utils";
import spacing from "../../../theme/spacing";

const dotColors = {
  white: getThemeValue("white", "base"),
  primary: getThemeValue("primary", "base"),
  dark: getThemeValue("darkFill")
};

const colorVariants = {
  standard: dotColors.white,
  standardDisabled: dotColors.white,
  special: dotColors.white,
  specialDisabled: dotColors.white,
  outline: dotColors.primary,
  outlineDisabled: dotColors.primary,
  transparent: dotColors.primary,
  transparentDisabled: dotColors.primary,
  outlineGray: dotColors.dark,
  outlineGrayDisabled: dotColors.primary
};

const wave = keyframes`
  0%, 60%, 100% {
    transform: initial;
  }

  30% {
    transform: translateY(-10px);
  }
`;

export const Dot = styled.span`
  display: inline-block;
  width: 6px;
  height: 6px;
  border-radius: 50%;
  background: ${({ variant }) => colorVariants[variant]};
  animation: ${wave} 0.8s linear infinite;

  &:nth-child(2) {
    animation-delay: -0.1s;
    margin: 0 ${spacing.cozy};
  }

  &:nth-child(3) {
    animation-delay: -0.5s;
  }
`;
Dot.displayName = "Dot";

export const LoaderWrapper = styled.span`
  position: relative;
`;
LoaderWrapper.displayName = "LoaderWrapper";

export const ContentWrapper = styled.span`
  visibility: ${({ isLoading }) => (isLoading ? "hidden" : "visible")};
`;
ContentWrapper.displayName = "ContentWrapper";

export const DotsWrapper = styled.span`
  position: absolute;
  white-space: nowrap;
  left: 0;
  right: 0;
`;
DotsWrapper.displayName = "DotsWrapper";
