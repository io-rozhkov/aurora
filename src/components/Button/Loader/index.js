import React from "react";
import PropTypes from "prop-types";

import { Dot, LoaderWrapper, ContentWrapper, DotsWrapper } from "./styles";
import { BUTTON_VARIANTS } from "../../constants";

const Loader = ({ children, loading, variant }) => (
  <LoaderWrapper>
    {loading && (
      <DotsWrapper data-bdd="loader">
        <Dot variant={variant} />
        <Dot variant={variant} />
        <Dot variant={variant} />
      </DotsWrapper>
    )}
    <ContentWrapper isLoading={loading}>{children}</ContentWrapper>
  </LoaderWrapper>
);

Loader.propTypes = {
  loading: PropTypes.bool,
  variant: PropTypes.oneOf(BUTTON_VARIANTS).isRequired,
  children: PropTypes.node.isRequired
};

Loader.defaultProps = {
  loading: false
};

export default Loader;
