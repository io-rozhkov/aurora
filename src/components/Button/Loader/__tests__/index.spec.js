import React from "react";
import renderer from "react-test-renderer";

import Loader from "..";

describe("Loader", () => {
  describe("when loading true & children", () => {
    let props;

    it("should match snapshot", () => {
      props = {
        children: "Children",
        loading: true
      };

      const component = renderer.create(<Loader {...props} />);
      const tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe("when loading false", () => {
    let props;

    it("should match snapshot", () => {
      props = {
        children: "Children"
      };

      const component = renderer.create(<Loader {...props} />);
      const tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
