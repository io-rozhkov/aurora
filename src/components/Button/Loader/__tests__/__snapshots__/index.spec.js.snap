// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader when loading false should match snapshot 1`] = `
.c0 {
  position: relative;
}

.c1 {
  visibility: visible;
}

<span
  className="c0"
>
  <span
    className="c1"
  >
    Children
  </span>
</span>
`;

exports[`Loader when loading true & children should match snapshot 1`] = `
.c2 {
  display: inline-block;
  width: 6px;
  height: 6px;
  border-radius: 50%;
  -webkit-animation: hZBpYk 0.8s linear infinite;
  animation: hZBpYk 0.8s linear infinite;
}

.c2:nth-child(2) {
  -webkit-animation-delay: -0.1s;
  animation-delay: -0.1s;
  margin: 0 8px;
}

.c2:nth-child(3) {
  -webkit-animation-delay: -0.5s;
  animation-delay: -0.5s;
}

.c0 {
  position: relative;
}

.c3 {
  visibility: hidden;
}

.c1 {
  position: absolute;
  white-space: nowrap;
  left: 0;
  right: 0;
}

<span
  className="c0"
>
  <span
    className="c1"
    data-bdd="loader"
  >
    <span
      className="c2"
    />
    <span
      className="c2"
    />
    <span
      className="c2"
    />
  </span>
  <span
    className="c3"
  >
    Children
  </span>
</span>
`;
